theano/tests/breakpoint.py

```diff
@@ -0,0 +1,140 @@
+                    Theano from applying certain optimizations to improve
+                    performance, reduce memory consumption and/or reduce
+                    numerical instability.
+
+            Detailed explanation:
+            As of 2014-12-01 the PdbBreakpoint op is not known by any
+            optimization. Setting a PdbBreakpoint op in the middle of a
+            pattern that is usually optimized out will block the optimization.
+
+    Example:
+
+    .. code-block:: python
+
+        import theano
+        import theano.tensor as T
+        from theano.tests.breakpoint import PdbBreakpoint
+
+        input = T.fvector()
+        target = T.fvector()
+
+        # Mean squared error between input and target
+        mse = (input - target) ** 2
+
+        # Conditional breakpoint to be activated if the total MSE is higher
+        # than 100. The breakpoint will monitor the inputs, targets as well
+        # as the individual error values
+        breakpointOp = PdbBreakpoint("MSE too high")
+        condition = T.gt(mse.sum(), 100)
+        mse, monitored_input, monitored_target = breakpointOp(condition, mse,
+                                                              input, target)
+
+        # Compile the theano function
+        fct = theano.function([input, target], mse)
+
+        # Use the function
+        print fct([10, 0], [10, 5]) # Will NOT activate the breakpoint
+        print fct([0, 0], [10, 5]) # Will activate the breakpoint
+
+
+    """
+
+    __props__ = ("name",)
+
+    def __init__(self, name):
+        self.name = name
+
+    def make_node(self, condition, *monitored_vars):
+
+        # Ensure that condition is a theano tensor
+        if not isinstance(condition, theano.Variable):
+            condition = theano.tensor.as_tensor_variable(condition)
+
+        # Validate that the condition is a scalar (else it is not obvious how
+        # is should be evaluated)
+        assert (condition.ndim == 0)
+
+        # Because the user might be tempted to instantiate PdbBreakpoint only
+        # once and apply it many times on different number of inputs, we must
+        # create a new instance of the op here, define the instance attributes
+        # (view_map and var_types) in that instance and then apply it on the
+        # inputs.
+        new_op = PdbBreakpoint(name=self.name)
+        new_op.view_map = {}
+        new_op.inp_types = []
+        for i in range(len(monitored_vars)):
+            # Every output i is a view of the input i+1 because of the input
+            # condition.
+            new_op.view_map[i] = [i+1]
+            new_op.inp_types.append(monitored_vars[i].type)
+
+        # Build the Apply node
+        inputs = [condition] + list(monitored_vars)
+        outputs = [inp.type() for inp in monitored_vars]
+        return Apply(op=new_op, inputs=inputs, outputs=outputs)
+
+    def perform(self, node, inputs, output_storage):
+        condition = inputs[0]
+        try:
```

Comment:
I would do that only in the if condition is True. Otherwise, it will transfer all data to the CPU at each execution.